### PR TITLE
Carry 11784: rmi dangling is unsafe when pulling

### LINKED ISSUE #8926: Docker rmi with dangling=true unsafe while images are being pulled
I was reading the docker documentation (https://docs.docker.com/reference/commandline/cli/), and noticed the following recommendation for batch clean up of dangling images:

```
sudo docker rmi $(sudo docker images -f "dangling=true" -q)
```

I was excited when I saw the dangling=true filter since having large numbers of dangling images was creating problems in our build environment. However, I believe this command is unsafe while pulling images since there there are brief points while images are being pulled when docker considers them dangling. To test this, I ran a script which simply pulls 10 images from our private registry, one after another. While this was running I ran the following command in a separate terminal:

```
while true; do docker rmi $(docker images -q -f "dangling=true"); sleep 1; done
```

I noticed that during the pull process, images were being occasionally deleted. Assuming the pull was successful, I will sometimes tag the image from `<private_registry>/<name>` to just `<name>`. After the end of the pull, all images are listed when I run `docker images`, however when I try to run an affected image, I see something like this:

```
stack@bc03:~$ docker run -ti cassandra
Unable to find image 'cassandra' locally
Pulling repository cassandra
2014/11/03 08:36:01 Error: image library/cassandra not found
```

The image cannot be found since it was retagged to remove the `<private_registry>` part of the name. My hope was that I could run this command automatically every day to clean up any dangling images, but that is not possible since I may accidentally destroy images that are being pulled.

I believe that the dangling=true filter should not show images or components of images that are in the process of being pulled...

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com
Signed-off-by: Tibor Vass tibor@docker.com

Fixes #8926

### CODE REVIEW

**Comment by icecrime** [comment]
Several thoughts:

* Code looks good on first look
* The issue was opened since many months
* I understand how hard it can be to test considering it's a race
* I'm afraid of the regression this kind introduce (if we miss a release call somewhere, we'll end up with a layer that can never be removed unless you restart your daemon).

So in the end I'm LGTM, but not for a 1.7.0 cherry-pick.

**Comment by tiborvass** [comment]
@icecrime fair enough, I'm removing the 1.7.0 milestone

**Comment by icecrime** [comment]
@tiborvass Can you please rebase?
@jlhawn @stevvooe @cpuguy83 Can you PTAL? It's simply a carry from #11784.

**Comment by cpuguy83** [comment]
LGTM but needs a rebase.

**Comment by tiborvass** [comment]
Rebased

**Comment by stevvooe** [comment]
@tiborvass @icecrime At this point, I've been through this PR several times. While this is a good fix, we need to better represent the lifecycle of a pull to manage the resources properly. Let's leave this open for now and merge if we don't have a solution for this before the 1.8 release.

**Comment by tiborvass** [comment]
@jlhawn if you could have a look that'd be great :)

File: builder/internals.go [hunk 1]
```diff
@@ -128,6 +128,8 @@ func (b *Builder) commit(id string, autoCmd *runconfig.Command, comment string)
 	if err != nil {
 		return err
 	}
+	b.Daemon.Graph().Retain(b.id, image.ID)
+	b.activeImages = append(b.activeImages, image.ID)
 	b.image = image.ID
```
Comment (jlhawn) [inline builder/internals.go]: There's still a race condition: the time between the `b.Daemon.Commit()` call above and the call to `b.Daemon.Graph().Retain()` here, though I guess there's really not much we can do about that.

**Comment by jlhawn** [comment]
LGTM if we're okay with the small race window on build commit.

File: daemon/image_delete.go [hunk 2]
```diff
@@ -137,6 +137,9 @@ func (daemon *Daemon) imgDeleteHelper(name string, list *[]types.ImageDelete, fi
 }
 
 func (daemon *Daemon) canDeleteImage(imgID string, force bool) error {
+	if daemon.Graph().IsHeld(imgID) {
```
Comment (stevvooe) [inline daemon/image_delete.go]: If the image gets held after this point, can it still be deleted? The issue with this test is that after it returns false, the condition may be true immediately after. You may need to hold a lock during this method to ensure a pull doesn't proceed until method has done its damage.

**Comment by stevvooe** [comment]
LGTM

I still have concerns for race conditions but no sense holding this back any longer.

**Comment by calavera** [comment]
cool, thank you guys, merging!

**Comment by dalanlan** [comment]
It might not be a good place for a question, yet i was wondering that whether the race condition issue has been fully addressed now? By which i mean, could `docke remove` hurts `docker pull` when it comes to the same image?

**Comment by cpuguy83** [comment]
@dalanlan Yes

**Comment by migueleliasweb** [comment]
Hey everyone, do this pull request also solves the problem with images being built while the remove dangling is running ?

**Comment by cpuguy83** [comment]
Yes

**Pull request merged**